From 102114fea355bc240777246ea46bfc2a58955768 Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Fri, 30 Sep 2022 12:39:19 +0300
Subject: [PATCH] [SW-69872] gaudi3: d2d: sync d2d init for both dies

In order to make sure both dies start initialization for d2d
link together, a sync mechanism is added.

This is because usually, slave will reach d2d init earlier than
master due to less processing compared to master.

Change-Id: Id5d6fa8ec9184c3d1e66dc5eb0434384135812d2
---

diff --git a/arch/gaudi3/arc/dev_common.h b/arch/gaudi3/arc/dev_common.h
index 7fd6e0f..f42e783 100644
--- a/arch/gaudi3/arc/dev_common.h
+++ b/arch/gaudi3/arc/dev_common.h
@@ -40,3 +40,4 @@
 int arcboot_rom_efuse_boot_info_skip(void);
 int arcboot_rom_fuse_pcie_ctrl_config(void);
 int arcboot_rom_fuse_d2d_spi_gw_en(void);
+void arcboot_rom_detect_signal_on_slave(void);
diff --git a/arch/gaudi3/arc/progress_bits.h b/arch/gaudi3/arc/progress_bits.h
index 0351a35..2d16bff 100644
--- a/arch/gaudi3/arc/progress_bits.h
+++ b/arch/gaudi3/arc/progress_bits.h
@@ -130,39 +130,40 @@
 	PROGRESS_ARCBOOT_ROM_PCIE_LINK_FINISH = 123,
 	PROGRESS_ARCBOOT_ROM_D2D_LINK_START = 124,
 	PROGRESS_ARCBOOT_ROM_D2D_LINK_STS_BUSY = 125,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_TIMEDOUT = 126,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_SUCCESS = 127,
-	PROGRESS_ARCBOOT_ROM_RECOVERY_END = 128,
-	PROGRESS_ARCBOOT_IN_EROM_RECOVERY = 129,
-	PROGRESS_INIT_EROM_BOOT_DESC_EXCEEDS_MAX_SIZE = 130,
-	PROGRESS_INIT_EROM_BOOT_DESC_DMA_START = 131,
-	PROGRESS_INIT_EROM_BOOT_DESC_CRC_FAIL = 132,
-	PROGRESS_INIT_EROM_BOOT_DESC_UNALIGNED_DEST = 133,
-	PROGRESS_INIT_EROM_BOOT_DESC_INVALID_IMG_SIZE = 134,
-	PROGRESS_INIT_EROM_BOOT_DESC_UNALIGNED_IMG_SIZE = 135,
-	PROGRESS_INIT_EROM_BOOT_DESC_DMA_ERROR = 136,
-	PROGRESS_INIT_EROM_BOOT_DESC_DMA_FINISH = 137,
-	PROGRESS_INIT_EROM_BOOT_DESC_HDR_MAGIC_FAIL = 138,
-	PROGRESS_INIT_EROM_BOOT_DESC_INVALID_IMG_OFFSET = 139,
-	PROGRESS_INIT_EROM_BOOT_DESC_VERIFICATION_FAIL = 140,
-	PROGRESS_INIT_EROM_BOOT_DESC_FAIL = 141,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_START = 142,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_VERSION_FAIL = 143,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_KEY_REVOKED = 144,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_KEY_HASH_FAIL = 145,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_AUTH_FAIL = 146,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_SUCCESS = 147,
-	PROGRESS_ARCBOOT_EROM_IMG0 = 148,
-	PROGRESS_ARCBOOT_EROM_IMG0_LOADADDR_INVALID = 149,
-	PROGRESS_ARCBOOT_EROM_IMG0_STARTADDR_INVALID = 150,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_START = 151,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_CRC_FAILED = 152,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_UNALIGNED_DEST = 153,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_INVALID_IMG_SIZE = 154,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_UNALIGNED_IMG_SIZE = 155,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_ERROR = 156,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_FINISH = 157,
-	PROGRESS_ARCBOOT_EROM_IMG0_VERIFICATION_FAIL = 158,
-	PROGRESS_ARCBOOT_EROM_IMG0_START = 159,
-	PROGRESS_ARCBOOT_EROM_IMG0_FINISH = 160,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_INTER_DIE_SYNC = 126,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_TIMEDOUT = 127,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_SUCCESS = 128,
+	PROGRESS_ARCBOOT_ROM_RECOVERY_END = 129,
+	PROGRESS_ARCBOOT_IN_EROM_RECOVERY = 130,
+	PROGRESS_INIT_EROM_BOOT_DESC_EXCEEDS_MAX_SIZE = 131,
+	PROGRESS_INIT_EROM_BOOT_DESC_DMA_START = 132,
+	PROGRESS_INIT_EROM_BOOT_DESC_CRC_FAIL = 133,
+	PROGRESS_INIT_EROM_BOOT_DESC_UNALIGNED_DEST = 134,
+	PROGRESS_INIT_EROM_BOOT_DESC_INVALID_IMG_SIZE = 135,
+	PROGRESS_INIT_EROM_BOOT_DESC_UNALIGNED_IMG_SIZE = 136,
+	PROGRESS_INIT_EROM_BOOT_DESC_DMA_ERROR = 137,
+	PROGRESS_INIT_EROM_BOOT_DESC_DMA_FINISH = 138,
+	PROGRESS_INIT_EROM_BOOT_DESC_HDR_MAGIC_FAIL = 139,
+	PROGRESS_INIT_EROM_BOOT_DESC_INVALID_IMG_OFFSET = 140,
+	PROGRESS_INIT_EROM_BOOT_DESC_VERIFICATION_FAIL = 141,
+	PROGRESS_INIT_EROM_BOOT_DESC_FAIL = 142,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_START = 143,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_VERSION_FAIL = 144,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_KEY_REVOKED = 145,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_KEY_HASH_FAIL = 146,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_AUTH_FAIL = 147,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_SUCCESS = 148,
+	PROGRESS_ARCBOOT_EROM_IMG0 = 149,
+	PROGRESS_ARCBOOT_EROM_IMG0_LOADADDR_INVALID = 150,
+	PROGRESS_ARCBOOT_EROM_IMG0_STARTADDR_INVALID = 151,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_START = 152,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_CRC_FAILED = 153,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_UNALIGNED_DEST = 154,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_INVALID_IMG_SIZE = 155,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_UNALIGNED_IMG_SIZE = 156,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_ERROR = 157,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_FINISH = 158,
+	PROGRESS_ARCBOOT_EROM_IMG0_VERIFICATION_FAIL = 159,
+	PROGRESS_ARCBOOT_EROM_IMG0_START = 160,
+	PROGRESS_ARCBOOT_EROM_IMG0_FINISH = 161,
 };
diff --git a/arch/gaudi3/arc/recovery.c b/arch/gaudi3/arc/recovery.c
index a27ae62..503de38 100644
--- a/arch/gaudi3/arc/recovery.c
+++ b/arch/gaudi3/arc/recovery.c
@@ -188,6 +188,16 @@
 	 */
 	arcboot_rom_d2d_phy_wd_hw_errata_fix();
 
+	/* Die 0 and Die 1 sync for d2d */
+	if (!arcboot_rom_global.die_id) {
+		/* Indicate slave die that d2d init has started */
+		arcboot_rom_notify_slave();
+	} else {
+		arcboot_rom_detect_signal_on_slave();
+	}
+	/* Inter die indication for d2d link init handshaking */
+	arcboot_rom_status_update(PROGRESS_ARCBOOT_ROM_D2D_LINK_INTER_DIE_SYNC);
+
 	/* Set START bit for D2D link initialization */
 	regmap_writel(D0_PSOC_GLOBAL_CONF_DPHY_FSM_CTRL_START_M, mmD0_PSOC_GLOBAL_CONF_DPHY_FSM_CTRL);
 
diff --git a/arch/gaudi3/arc/slave.c b/arch/gaudi3/arc/slave.c
index 095e8ef..6118f15 100644
--- a/arch/gaudi3/arc/slave.c
+++ b/arch/gaudi3/arc/slave.c
@@ -24,20 +24,40 @@
 	regmap_writel(val, mmD0_PSOC_GLOBAL_CONF_DIE_RECOVERY_OUT);
 }
 
-void arcboot_rom_wait_for_signal(void)
+void arcboot_rom_detect_signal_on_slave(void)
 {
 	uint32_t val;
 
+	while (1) {
+		val = regmap_readl(mmD0_PSOC_GLOBAL_CONF_DIE_RECOVERY_IN);
+		if (val & D0_PSOC_GLOBAL_CONF_DIE_RECOVERY_IN_IND_M) {
+			break;
+		}
+	}
+	/*
+	* No timeout added for while(1) due to the usage of this function:
+	* 1) Recovery: It is required for slave to wait for this indication from
+	* master. As slave die in rom is to only execute recovery, it waits forever
+	* for this indication to be set, apart from recovery there is no logic to be
+	* executed on slave in ROM.
+	*
+	* 2) D2D link init: The pcie linkup time on master cannot be predicted, hence
+	* slave needs to wait. However the implication of this while(1) is minimal as
+	* d2d is the last process in the recovery flow which ends also in a while(1).
+	* Hence this would not have any functional impact.
+	*/
+}
+
+void arcboot_rom_wait_for_signal(void)
+{
 	if (!arcboot_rom_global.die_id)
 		return;
 
 	arcboot_rom_status_update(PROGRESS_SLV_DIE_WAITING_ON_RECOVERY_IND);
 
-	while (1) {
-		val = regmap_readl(mmD0_PSOC_GLOBAL_CONF_DIE_RECOVERY_IN);
-		if (val & D0_PSOC_GLOBAL_CONF_DIE_RECOVERY_IN_IND_M)
-			execute_recovery();
-	}
+	arcboot_rom_detect_signal_on_slave();
+
+	execute_recovery();
 }
 
 #define FUSE_EMULATION_DATA_UDELAY	5000 /* 5ms */
